### PR TITLE
KAFKA-14183; Cluster metadata bootstrap file should use header/footer

### PR DESCRIPTION
**Pull request by jsancio**
The boostrap.checkpoint files should include a control record batch for
the SnapshotHeaderRecord at the start of the file. It should also
include a control record batch for the SnapshotFooterRecord at the end
of the file.

The snapshot header record is important because it versions the rest of
the bootstrap file.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: metadata/src/main/java/org/apache/kafka/metadata/util/BatchFileWriter.java [hunk 1]
```diff
@@ -36,16 +39,24 @@
 
 
 /**
- * Write an arbitrary set of metadata records into a Kafka metadata log batch format. This is similar to the binary
- * format used for metadata snapshot files, but the log epoch and initial offset are set to zero.
+ * Write an arbitrary set of metadata records into a Kafka metadata log batch format.
+ *
+ * This is similar to the binary format used for metadata snapshot files, but the log epoch
+ * and initial offset are set to zero.
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/metadata/util/BatchFileWriter.java]: We should document that this class adds the two control records automatically

File: raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java [hunk 2]
```diff
@@ -260,18 +260,18 @@ private void appendControlMessage(Function<ByteBuffer, MemoryRecords> valueCreat
     /**
      * Append a {@link LeaderChangeMessage} record to the batch
      *
-     * @param @LeaderChangeMessage The message to append
-     * @param @currentTimeMs The timestamp of message generation
+     * @param LeaderChangeMessage The message to append
+     * @param currentTimestamp The current time
```
Comment (mumrah) [inline raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java]: nit "current time in millis" ?

**Review by mumrah — APPROVED** [review]
@jsancio thanks for the patch! A few minor non-blocking comments inline. I filed a follow-up to add unit tests for BatchFileWrite since I realize I never included any (not needed for this PR). https://issues.apache.org/jira/browse/KAFKA-14186

LGTM

**Comment by jsancio** [comment]
> @jsancio thanks for the patch! A few minor non-blocking comments inline. I filed a follow-up to add unit tests for BatchFileWrite since I realize I never included any (not needed for this PR). https://issues.apache.org/jira/browse/KAFKA-14186

@mumrah, Yes, we could use more tests for this type. I added BatchFileWriterReaderTest in this commit to test basic header and footer support.

**Comment by jsancio** [comment]
Unrelated test failure: 

>Build / JDK 11 and Scala 2.13 / kafka.network.DynamicConnectionQuotaTest.testDynamicListenerConnectionCreationRateQuota()

**Pull request merged**